Snowflake.Data/Core/SFDataConverter.cs

```diff
@@ -205,8 +205,7 @@ private static DateTimeOffset ConvertToDateTimeOffset(string srcVal, SFDataType
                     }
                     else
                     {
-                        long millis = (long)((DateTime)srcVal).ToUniversalTime().Subtract(

```

Comment:
It's probably a good idea to change the behavior. We can change the test case to specify the input as UTC time with a Z.

Reply:
Not really sure what you mean by this. 

Please note that the problem is not in the test case. I dont think you should change the test case.
The issue is that if your client computer is in a timezone after UTC and you send in a normal datetime (with only a date and no timezone specified), the wrong date will be stored in snowflake.
As I said, this can be fixed simply by removing the call to ToUniversalTime()

It is not easy to write a unit test for this scenario since there really is no good way to change the timezone programmatically. The time zone setting in windows is system-wide - it can not be changed per thread or per process. But you can easily recreate the scenario by manually changing the timezone of your development computer to UTC+1 before running the unit tests.

Reply:
I suggest that you modify csharpTypeValToSfTypeVal to actually return a Date only, so regardless of the time part specified and the kind of Datetime you should always return a date only. 

This can be done with the following change:

`                case DbType.Date:
                    destType = SFDataType.DATE.ToString();
                    if (srcVal.GetType() != typeof(DateTime))
                    {
                        throw new SnowflakeDbException(SFError.INVALID_DATA_CONVERSION, srcVal, 
                            srcVal.GetType().ToString(), DbType.Date.ToString());
                    }
                    else
                    {
                        DateTime dt = ((DateTime)srcVal).Date;
                        var ts = dt.Subtract(UnixEpoch);
                        long millis = (long)(ts.TotalMilliseconds);
                        destVal = millis.ToString();
                    }
                    break;
`

I also suggest adding the following test case to make sure it works regardless of the kind of datetime used and the timezone the computer is in:

`        [Test]
        [TestCase("2100-12-31 23:59:59.9999999", DateTimeKind.Utc)]
        [TestCase("2100-12-31 23:59:59.9999999", DateTimeKind.Local)]
        [TestCase("2100-12-31 23:59:59.9999999", DateTimeKind.Unspecified)]
        [TestCase("2200-01-01 00:00:00.0000000", DateTimeKind.Utc)]
        [TestCase("2200-01-01 00:00:00.0000000", DateTimeKind.Local)]
        [TestCase("2200-01-01 00:00:00.0000000", DateTimeKind.Unspecified)]
        [TestCase("1960-01-01 00:00:00.0000000", DateTimeKind.Unspecified)]
        [TestCase("9999-12-31 23:59:59.9999999", DateTimeKind.Unspecified)]
        [TestCase("1982-01-18 16:20:00.6666666", DateTimeKind.Unspecified)]
        [TestCase("1982-01-18 23:59:59.0000000", DateTimeKind.Unspecified)]
        [TestCase(null, DateTimeKind.Unspecified)]
        public void TestConvertDate(string inputTimeStr, object kind=null)
        {
            if (kind == null)
                kind = 0;
            DateTime inputTime;
            if (inputTimeStr == null)
            {
                inputTime = DateTime.Now;
            }
            else
            {
                inputTime = DateTime.ParseExact(inputTimeStr, "yyyy-MM-dd HH:mm:ss.fffffff", CultureInfo.InvariantCulture);
            }
            var dtExpected = inputTime.Date;
            internalTestConvertDate(dtExpected, DateTime.SpecifyKind(inputTime, (DateTimeKind)kind));
        }

        private void internalTestConvertDate(DateTime dtExpected, DateTime testValue)
        {
            var result = SFDataConverter.csharpTypeValToSfTypeVal(System.Data.DbType.Date, testValue);
            // Convert result to DateTime for easier interpretation
            var unixEpoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
            DateTime dtResult = unixEpoch.AddMilliseconds(Int64.Parse(result.Item2));
            Assert.AreEqual(dtExpected, dtResult);
        }
`